### PR TITLE
Silence dimensions with noisy ML models

### PR DESCRIPTION
**Pull request by vkalintiris**
##### Summary

See https://github.com/netdata/netdata/discussions/14993 for more info.

A couple notes:

- We silence dimensions, not models.
- A dimension stops being silenced whenever we fulfill a training request (regardless of whether it is trained successfully or not).
- You can track the number of silenced dimensions in the `netdata.training_status_on_{hostname}` chart.

##### Test Plan

- CI jobs,
- Manual testing on staging

### CODE REVIEW

**Comment by andrewm4894** [comment]
@vkalintiris 

> You can track the number of silenced dimensions in the netdata.training_status_on_{hostname} chart.

will this be a new dim on this chart?

**Comment by andrewm4894** [comment]
think i'll need to loosen the thresholds a bit to test that it behaves as expected - i'll do that in an updated config on my end tomorrow.

File: ml/Config.cc
Comment (stelfrag) [inline ml/Config.cc:50]: In the example of the linked issue, this can be a float number

**Comment by andrewm4894** [comment]
This seems to be working as expected. 

Here is a chart of all silenced dims over last 24 hours on a node i was testing the defaults on. 

image

As we can see it's not actually "silencing" that many dims all that often.

So its a bit hard to test or at least fully say what the value or impact here is since its more like an insurance feature against dims getting excessively stuck in bad state. 

I think we should merge it as is with the default params as is so that its set to be very generous and not really going to kick in unless really bad outcome of dim firing for over 50% within a 1800 window.

In some sense, this essentially puts a cap on how long the anomaly rates can fire but i think given the defaults as so generous we are ok to not really worry about this as being anything that could hurt things like node level anomaly rate performance etc in terms of its use and value.

**Comment by andrewm4894** [comment]
I will update the docs in a follow on PR once this is merged.

**Comment by andrewm4894** [comment]
I am going to just do one last test of updating the above agent to have below config and see impact on the silenced dim over next few hours. 

```
[ml]
    enabled = yes
    maximum num samples to train = 14400
    train every = 3600
    number of models per dimension = 1
    enable statistics charts = yes
    dimension anomaly rate suppression window = 900
    dimension anomaly rate suppression threshold = 0.5
```

**Comment by vkalintiris** [comment]
> I am going to just do one last test of updating the above agent to have below config and see impact on the silenced dim over next few hours.
> 
> ```
> [ml]
>     enabled = yes
>     maximum num samples to train = 14400
>     train every = 3600
>     number of models per dimension = 1
>     enable statistics charts = yes
>     dimension anomaly rate suppression window = 900
>     dimension anomaly rate suppression threshold = 0.5
> ```

The last option should always be expressed with number of samples, so instead of:

```
dimension anomaly rate suppression threshold = 0.5
```

It should be:

```
dimension anomaly rate suppression threshold = 450
```

**Review by andrewm4894 — APPROVED** [review]
working as expected and i think a nice sort of worse case or upper bound type insurance type of feature

**Review by stelfrag — APPROVED** [review]
Approving as it is now -- if the `suppression_anomaly_transition_counter` is also addressed I will review again. On a "normal" system it is indeed hard to get silenced dimensions

**Comment by vkalintiris** [comment]
> And then a dim would be supressed if its on for 900/1800 AND also has at least 450 transitions from 0->1 or 1->0

Not a fan of adding more state to the dimensions. I get why you are suggesting this and it makes sense. However, I'd like us to add functionality only if we have examples where the current suppression logic is suppressing dimensions that shouldn't be suppressed (and the number of such suppressed dimensions is significant enough that it hinders one's ability to figure out what happened when an anomaly occurs).

**Comment by andrewm4894** [comment]
@ktsaou fyi I think something like this could be a easy quick win to try calm down and silence noisey dims.

https://github.com/netdata/netdata/pull/15011#issuecomment-1540802125

Might need a bit more thoughts but just sharing as fyi.

**Pull request merged**
